### PR TITLE
Re-enable `rustc_codegen_gcc` tests in CI

### PR DESCRIPTION
**Pull request by cuviper**
When #117947 dropped llvm-15 from CI, we neglected to copy #117313's changes to enable `rustc_codegen_gcc` testing to the new base llvm-16. This is now restored, as well as copying the setup to llvm-17 as well so we hopefully won't miss it next time.

In addition, due to case mismatch in `$extra_env` updates in `docker/run.sh`, I think it wasn't actually getting enabled before, but this should now be fixed. I also avoided the linker hack for `libgccjit.so` that was present before, because that's not needed if the version matches the base `gcc` used for linking.

r? GuillaumeGomez

### CODE REVIEW

**Comment by cuviper** [comment]
... at least we know the tests really did run this time! :sweat_smile:

**Comment by GuillaumeGomez** [comment]
Ah. Well, I'll try to fix it tomorrow. :sweat_smile:

File: src/ci/docker/run.sh [hunk 1]
```diff
@@ -269,14 +269,14 @@ touch $objdir/${SUMMARY_FILE}
 
 extra_env=""
 if [ "$ENABLE_GCC_CODEGEN" = "1" ]; then
-  extra_env="$EXTRA_ENV --env ENABLE_GCC_CODEGEN=1"
+  extra_env="$extra_env --env ENABLE_GCC_CODEGEN=1"
```
Comment (antoyo) [inline src/ci/docker/run.sh:272]: `extra_env` is set to `""` just before this condition.

Comment (cuviper) [reply src/ci/docker/run.sh:272]: Yeah, ok, that one doesn't really matter, but each following line was clobbering the ones before.

File: src/ci/docker/run.sh [hunk 2]
```diff
@@ -269,7 +269,7 @@ touch $objdir/${SUMMARY_FILE}
 
 extra_env=""
 if [ "$ENABLE_GCC_CODEGEN" = "1" ]; then
-  extra_env="$extra_env --env ENABLE_GCC_CODEGEN=1"
+  extra_env="$EXTRA_ENV --env ENABLE_GCC_CODEGEN=1"
```
Comment (cuviper) [inline src/ci/docker/run.sh:272]: I don't understand why you put this back -- Is there an `$EXTRA_ENV` coming from somewhere else?

Comment (GuillaumeGomez) [reply src/ci/docker/run.sh:272]: I was just confirming that it was not that. The problem currently is that we use the old mangling version somewhere and I'm trying to understand where. I'll remove/replace this commit often I think. Sorry for the noise. ^^'

**Pull request merged**